text/0000-wallet-replay-protection.md

```diff
@@ -0,0 +1,298 @@
+  PS = generate_protected_outputs()
+  return [anchor] + PS
+}
+
+// A protected version of a receive transaction. Returns a list of inputs and a list of outputs
+// In case an anchor does not exist, it contributes an anchor and PS outputs. If an anchor already
+// exists, it is a PayJoin transaction - we add a protected output as an input
+fn receive_protected() -> ([]Output, []Output) {
+  // if we have no anchor, contribute anchor + PS outputs for replay protection
+  // TODO: this might be too obvious from the chain analysis perspective
+  if anchor_not_exists:
+    return [], bootstrap_outputs()
+  // PayJoin - protect the transaction with a protect output as an input
+  return [random_protected_output()], []
+}
+
+// Receive txs don't need an absolute protection so we flip a coin to determine whether
+// it will be a regular or protected transaction based on the wallet configuration
+fn receive(value: int64) -> ([]Output, []Output) {
+  receiver_inputs = []
+  receiver_outputs = []
+  // receive-only wallets could have receive_protected_prob=0.0 to always have 1-2 receive transactions
+  is_protected = random() <= config.RECEIVE_PROTECTED_PROB
+  if is_protected {
+    receiver_inputs, receiver_outputs = receive_protected()
+    receiver_outputs += [generate_protected_output(value)]
+  } else {
+    receiver_outputs += [generate_unprotected_output(value)]
+  }
+  return receiver_inputs, receiver_outputs
+}
+
+// A 'send' transaction needs to always be protected from replays so we MUST
+// either add a protected output as an input or make it a bootstrap transaction
+// if an anchor does not exist yet
+fn send(value: int64) -> ([]Output, []Output) {
+  // first pick the inputs that hold enough value to perform a tx
+  inputs = pick_random_inputs()
+  outputs = []
+
+  // if we have no anchor, contribute anchor + PS outputs for replay protection
+  if anchor_not_exists() {
+    outputs = bootstrap_outputs()
+  // otherwise, add a protected output as input to protect the transaction
+  } else {
+    // contribute a protected output as an input if it's not already contributed
+    if count_protected_output(inputs) == 0 {
+        inputs += [random_protected_output()]
+    }
+  }
+  change_value = sum(input.v for input in inputs) - value  // fee is missing from here
+  // we know our change output will be protected either by bootstrap outputs or by
+  // a protected input so we label it as a protected output
+  change_output = generate_protected_output(change_value)
+  outputs += [change_output]
+  return inputs, outputs
+}
+
+// TODO: needs handling of running out of protected outputs and keeping a minimal of `N`
+// protected outputs as defined in the configuration file
+```
+
+TODO: Check this is safe.
+
+#### Protection with wallet history
+
+A wallet can remember which outputs were spent. This way, if a spent output reappears in the wallet, a user is given a choice to either accept it or refresh it through a self-spend transaction.
+

```

Comment:
Perhaps a previously-spent-utxo should only show up if a user asks to see them.
We don't want to bother non-expert users with them.
Obviously they would never contribute to the wallet balance.

Reply:
I'm thinking that if we label outputs in the bulletproof (both protected _and_ unprotected) then we have 3 states.
1. `old` outputs that don't have labeling in bulletproofs
2. `protected` outputs
3. `unprotected` outputs

Suppose we create an `anchor` output and we start labeling all outputs we create by either giving them protected or unprotected label in the bulletproof bytes.

If we have an `anchor` created and we receive an output that is `old`, then it must have come from a replay of one of the old transactions, correct? An exception to this would be if someone perfectly timed a wallet restore in which case we wouldn't know which outputs were recently received, but it might work for recognizing past replay attacks and we might be able to recognize and ignore these (or as you said, ask the user to show them). Might even have a metric for these events sent to a public monitoring system to observe the number of such cases.

Reply:
To repeat an observation elsewhere, old outputs that don't have a labeling in bulletproofs should automatically be classified as unprotected. This can be achieved with either a new derivation path or a 1 flag bit instead of a current 0 flag bit.

Reply:
We can use a single flag bit to distinguish between 4 types of output: Decoys are 0-valued 0-flagged, anchors are 0-valued 1-flagged, unprotected are positive-valued 0-flagged, and protected are positive-valued 1-flagged.

Reply:
decoy could actually be protected (depends a lot on how we create decoys e.g. if we add an input along with it). Would we make sure that a decoy output is never protected or would we label it as unprotected regardless?

Reply:
The wallet could in principle mark specific decoys as protected, and use them like other protected outputs. But without using additional flag bits, this knowledge would not survive a restore.
After restore, it would appear the same as any other decoy, and must be presumed unprotected.

I think that a wallet that always considers them as unprotected is slightly simpler.
The question is how a wallet would create them in the first place? That will probably determine how best to spend them. For spending them as actual decoys, being protected may not matter.

Reply:
Indeed. Decoys seem to be a whole new topic to discuss separately from this RFC though